include/picongpu/initialization/ParserGridDistribution.hpp

```diff
@@ -77,40 +90,57 @@ class ParserGridDistribution
 
     /** Get local Size of this dimension
      *
-     *  \param[in] gpuPos as integer in the range [0, n-1] for this dimension
+     *  \param[in] devicePos as unsigned integer in the range [0, n-1] for this dimension
      *  \return uint32_t with local number of cells
      */
     uint32_t
-    getLocalSize( const int gpuPos ) const
+    getLocalSize( uint32_t const devicePos ) const
     {
         value_type::const_iterator iter = parsedInput.begin();
-        // go to last gpu of this block b{n}
-        int i = iter->second - 1;
+        // go to last device of these n subdomains extent{n}
+        uint32_t i = iter->count - 1u;
 
-        while( i < gpuPos )
+        while( i < devicePos )
         {
             ++iter;
-            // go to last gpu of this block b{n}
-            i += iter->second;
+            // go to last device of these n subdomains extent{n}
+            i += iter->count;
         }
 
-        return iter->first;
+        return iter->extent;
+    }
+
+    /** Verify the number of subdomains matches the devices
+     *
+     * Check that the number of subdomains in a dimension, after we
+     * expanded all regexes, matches the number of devices for it.
+     *
+     * \param[in] numDevices number of devices for this dimension
+     */
+    void
+    verifyDevices( uint32_t const numDevices ) const
+    {
+        uint32_t numSubdomains = 0u;
+        for( SubdomainPair const & p : parsedInput )
+            numSubdomains += p.count;
+
+        PMACC_VERIFY( numSubdomains == numDevices );
     }
 
 private:
     value_type parsedInput;
 
-    /** Parses the input string to a vector of pairs
+    /** Parses the input string to a vector of SubdomainPair(s)
      *
-     *  Parses the input string in the form a,b,c{n},d{m},e,f
-     *  to a vector of pairs with base number (a,b,c,d,e,f) and multipliers
-     *  (1,1,n,m,e,f)
+     * Parses the input string in the form a,b,c{n},d{m},e,f
+     * to a vector of SubdomainPair with extent number (a,b,c,d,e,f) and
+     * counts (1,1,n,m,e,f)
      *
-     *  \param[in] s as const std::string in the form a,b{n}
-     *  \return std::vector<pair> with uint32_t (base, multiplier)
+     * \param[in] s as string in the form a,b{n}
+     * \return std::vector<SubdomainPair> with 2x uint32_t (extent, count)
      */
     void
-    parseString( const std::string s )
+    parseString( std::string const s )
```

Comment:
Not important, but easy to implement: 

- returning a `value_type` instead of writing the result to directly to `parsedInput` seems more natural in terms of how it is used in the constructor (however, then the weird `value_type` might also have to be renamed to smth like `SubdomainPairs`)
- `parseString` could be safely shortened to `parse` given the argument type;

Reply:
good point, refactored that way now. I keep the `value_type` for now, it doesn't harm too much.